src/hooks/useForm.ts

```diff
@@ -0,0 +1,192 @@
+
+  // 필드 값 설정
+  const setValue = useCallback(<K extends keyof T>(field: K, value: T[K]) => {
+    setFields((prev) => ({
+      ...prev,
+      [field]: {
+        ...prev[field],
+        value,
+      },
+    }));
+  }, []);
+
+  // 에러 설정
+  const setError = useCallback(
+    <K extends keyof T>(field: K, error: string | null) => {
+      setFields((prev) => ({
+        ...prev,
+        [field]: {
+          ...prev[field],
+          error,
+        },
+      }));
+    },
+    []
+  );
+
+  // touched 상태 설정
+  const setTouched = useCallback(
+    <K extends keyof T>(field: K, touched: boolean = true) => {
+      setFields((prev) => ({
+        ...prev,
+        [field]: {
+          ...prev[field],
+          touched,
+        },
+      }));
+    },
+    []
+  );
+
+  // 특정 필드 유효성 검사
+  const validateField = useCallback(
+    <K extends keyof T>(field: K) => {
+      const fieldConfig = config[field];
+      const value = fields[field].value;
+
+      let error: string | null = null;
+
+      // required 검사
+      if (fieldConfig.required && !value) {
+        error = '필수 입력 항목입니다.';
+      }
+      // 커스텀 validator 검사
+      else if (fieldConfig.validator) {
+        error = fieldConfig.validator(value);
+      }
+
+      setError(field, error);
+      return error === null;
+    },
+    [fields, config, setError]
+  );
+
+  // 모든 필드 유효성 검사
+  const validateAll = useCallback(() => {
+    let isValid = true;
+    Object.keys(config).forEach((key) => {
+      const fieldValid = validateField(key as keyof T);
+      if (!fieldValid) {
+        isValid = false;
+      }
+    });
+    return isValid;
+  }, [config, validateField]);
+
+  // 폼 초기화
+  const reset = useCallback(() => {
+    setFields(createInitialState());
+  }, []);
+
+  // 필드 변경 핸들러
+  const handleChange = useCallback(
+    <K extends keyof T>(field: K) =>
+      (value: T[K]) => {
+        setValue(field, value);
+        // 이미 touched된 필드는 실시간 유효성 검사
+        if (fields[field].touched) {
+          validateField(field);
+        }
+      },
+    [fields, setValue, validateField]
+  );
+
+  // 필드 blur 핸들러
+  const handleBlur = useCallback(
+    <K extends keyof T>(field: K) =>
+      () => {
+        setTouched(field, true);
+        validateField(field);
+      },
+    [setTouched, validateField]
+  );
+
+  // 현재 값들 추출
+  const values = Object.keys(fields).reduce((acc, key) => {
+    acc[key as keyof T] = fields[key as keyof T].value;
+    return acc;
+  }, {} as T);
+
+  // 현재 에러들 추출
+  const errors = Object.keys(fields).reduce(
+    (acc, key) => {
+      acc[key as keyof T] = fields[key as keyof T].error;
+      return acc;
+    },
+    {} as { [K in keyof T]: string | null }
+  );
+
+  // 전체 폼 유효성 상태
+  const isValid =
+    Object.values(errors).every((error) => error === null) &&
+    Object.keys(config).every((key) => {
+      const fieldConfig = config[key];
+      const value = fields[key as keyof T].value;
+      return !fieldConfig.required || !!value;
+    });
```

Comment:
가독성을 위해 긴 조건들은 변수로 추상화하여 변수명으로 의미를 드러내도록 하면 유지보수에 좋을 것 같아요

예시
```suggestion
  const isValid = !hasErrors || !missingRequired;
```